src/llmcompressor/modifiers/quantization/calibration.py

```diff
@@ -0,0 +1,251 @@
+import logging
+
+import torch
+from compressed_tensors.quantization import QuantizationStatus, is_attention_module
+from compressed_tensors.quantization.lifecycle.forward import forward_quantize
+from compressed_tensors.quantization.utils import is_kv_cache_quant_scheme
+from compressed_tensors.utils.offload import is_module_offloaded, update_parameter_data
+from torch.nn import Module
+
+from llmcompressor.modifiers.quantization.cache import QuantizedKVParameterCache
+from llmcompressor.observers import Observer
+
+__all__ = [
+    "initialize_observer",
+    "update_weight_zp_scale",
+    "calibrate_input_hook",
+    "calibrate_output_hook",
+    "calibrate_kv_cache_input_hook",
+    "calibrate_kv_cache_output_hook",
+    "set_unset_kv_cache",
+    "freeze_module_quantization",
+]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def initialize_observer(
+    module: Module,
+    base_name: str,
+):
+    """
+    Initialize observer module and attach as submodule.
+    The name of the observer is fetched from the quantization_args.
+    The name is then used to load the observer from the registry and attached
+    to the module. The name of the observer uses the base_name provided.
+
+    :param module: torch.nn.Module that the observer is being attached to
+    :param base_name: str used to name the observer attribute
+
+    """
+
+    arg_name = "weights" if base_name == "weight" else f"{base_name}_activations"
+    quantization_scheme = getattr(module, "quantization_scheme", None)
+    if not quantization_scheme:
+        # no quantization scheme nothing to do
+        return
+
+    # observers have a different lifecycle for kv_cache
+    if is_attention_module(module):
+        return
+
+    quantization_args = getattr(quantization_scheme, arg_name, None)
+    # dont need observers for dynamic
+    if quantization_args and not quantization_args.dynamic:
+        observer = quantization_args.get_observer()
+        observer = Observer.load_from_registry(
+            observer, quantization_args=quantization_args
+        )
+        module.register_module(f"{base_name}_observer", observer)
+
+
+def call_observer(module: Module, base_name: str, value: torch.Tensor):
```

Comment:
This function is called for activations, however it still uses the weight's g_idx in call cases

Reply:
that's irrelevant to the core purpose of this function, which is to call the observer.

Reply:
If the observer being called is an activation observer, won't it attempt to permute the activation according to the weight_g_idx?

Reply:
We dont support grouped quant for activation ordering, only per tensor or per token, so that scenario would not happen but I agree we it's unclear from the current interface

Reply:
I see, I still recommend moving line 74 to line 76 just so we don't run into this issue if we do want to support group-wise activation quantization

Reply:
Updated to conditionally pass in g_idx